### PR TITLE
More useful S4 class

### PR DESCRIPTION
**Pull request by flying-sheep**
S4 classes currently expose too many externals and are hard to create from python (as you need to create and pass a lexeme)

This implements my suggestions from #14

### CODE REVIEW

**Comment by flying-sheep** [comment]
huh, `setup.py` [says](https://github.com/ralhei/pyRserve/blob/916b14843f20783a006e2a0259fc43bce7d300c9/setup.py#L40) this is python 3 only, but the tests run on python 2?

either the tests have to be switched to python 3 or python 2 has to be added to the trove classifiers

**Comment by ralhei** [comment]
You are right, the trove classifier is wrong, it should be 'python 2' for now, the Python3 implementation doesn't work yet. I'll take care of it asap. For now could you please adapt your code to support python2 (and 3)? I'll make suggestions in the code review.

File: pyRserve/rparser.py [hunk 1]
```diff
@@ -645,10 +656,17 @@ def __repr__(self):
         return '<Closure instance %d>' % id(self)
 
 
-class S4(object):
+class S4(dict):
     """Very simple representation of a S4 instance"""
-    def __init__(self, lexeme):
-        self.lexeme = lexeme
+    def __init__(self, source=(), **entries):
+        super().__init__(source, **entries)
```
Comment (ralhei) [inline pyRserve/rparser.py]: Should be
        super(S4, self).__init__(source, **entries)
for Py2

File: pyRserve/rparser.py [hunk 2]
```diff
@@ -645,10 +656,17 @@ def __repr__(self):
...
+
+        if 'class' in self:
+            self.classes = self['class']
+            del self['class']
+        else:
+            self.classes = []
 
     def __repr__(self):
-        return '<S4 attrs=%s>' % repr(self.lexeme.attr)
+        attrs = super().__repr__()
```
Comment (ralhei) [inline pyRserve/rparser.py]: Should be
`        attrs = super(S4, self).__repr__()`
for Py2

File: testing/test_rparser.py [hunk 3]
```diff
@@ -471,6 +471,37 @@ def test_eval_very_convoluted_function_result():
...
+    S4 classes behave like dicts but usually have a 'class' attribute.
+    """
+    res = conn.r('''
+        track <- setClass("track",
+                          slots = c(x="numeric", y="NULL"))
+        track(x = 1:10, y = NULL)
+    ''')
+    assert isinstance(res, rparser.S4)
+    assert res.classes == ['track']
+    assert res.keys() == {'x', 'y'}
```
Comment (ralhei) [inline testing/test_rparser.py]: Should be 
`    assert set(res.keys()) == {'x', 'y'}`
for Py2

File: testing/test_rparser.py [hunk 4]
```diff
@@ -471,6 +471,37 @@ def test_eval_very_convoluted_function_result():
...
+    assert compareArrays(res['x'], numpy.arange(1, 11))
+    assert res['y'] is None
+    assert "<S4 classes=['track'] {" in repr(res)
+
```
Comment (ralhei) [inline testing/test_rparser.py]: Please add extra empty line for pep8 compliance

Comment (flying-sheep) [inline testing/test_rparser.py]: sure! you might consider adding “pycodestyle” (formerly “pep8”) to the tests

**Comment by flying-sheep** [comment]
OK, all done!

**Comment by ralhei** [comment]
I saw #16, and I'm about finalizing for Python3. There has just been reported an issue #17  with Python3 which I can reproduce on my side, seems to fail on py 3.5.1 and py 3.6.1, don't know why it works for you ;-) But I'm going to fix this and push the change today.

**Pull request merged**